src/linker/Linker/DgmlDependencyRecorder.cs

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Compression;
+using System.Xml;
+using Mono.Cecil;
+
+namespace Mono.Linker
+{
+	/// <summary>
+	/// Class which implements IDependencyRecorder and writes the dependencies into an DGML file.
+	/// </summary>
+	public class DgmlDependencyRecorder : IDependencyRecorder, IDisposable
+	{
+		public const string DefaultDependenciesFileName = "linker-dependencies.dgml.gz";
+
+
+		private readonly LinkContext context;
+		private XmlWriter? writer;
+		private Stream? stream;
+
+		public DgmlDependencyRecorder (LinkContext context, string? fileName = null)
+		{
+			this.context = context;
+
+			XmlWriterSettings settings = new XmlWriterSettings {
+				Indent = true,
+				IndentChars = " "
+			};
+
+			if (fileName == null)
+				fileName = DefaultDependenciesFileName;
+
+			if (string.IsNullOrEmpty (Path.GetDirectoryName (fileName)) && !string.IsNullOrEmpty (context.OutputDirectory)) {
+				fileName = Path.Combine (context.OutputDirectory, fileName);
+				Directory.CreateDirectory (context.OutputDirectory);
+			}
+
+			var depsFile = File.OpenWrite (fileName);
+
+			if (Path.GetExtension (fileName) == ".dgml")
+				stream = depsFile;
+			else
+				stream = new GZipStream (depsFile, CompressionMode.Compress);
+
+			writer = XmlWriter.Create (stream, settings);
+			writer.WriteStartDocument ();
+			writer.WriteStartElement ("DirectedGraph", "http://schemas.microsoft.com/vs/2009/dgml");
+		}
+
+		public void WriteDgmlGraphToFile ()
+		{
+			Debug.Assert (writer != null);
+
+			writer.WriteStartElement ("Nodes");
+			{
+				foreach (var pair in nodeList) {
+					writer.WriteStartElement ("Node");
+					writer.WriteAttributeString ("Id", pair.Value.ToString ());
+					writer.WriteAttributeString ("Label", pair.Key);
+					writer.WriteEndElement ();
+				}
+			}
+			writer.WriteEndElement ();
+
+			writer.WriteStartElement ("Links");
+			{
+				foreach (var tup in linkList) {
+					writer.WriteStartElement ("Link");
+					writer.WriteAttributeString ("Source", nodeList[tup.dependent].ToString ());
+					writer.WriteAttributeString ("Target", nodeList[tup.dependee].ToString ());
+					writer.WriteAttributeString ("Reason", tup.reason);
+					writer.WriteEndElement ();
+				}
+			}
+			writer.WriteEndElement ();
+		}
+
+		public void Dispose ()
+		{
+			if (writer == null)
+				return;
+
+			// put all code to generate dgml here
+			WriteDgmlGraphToFile ();
+
+			writer.WriteStartElement ("Properties");
```

Comment:
Seems weird pattern to me to mandatory data during dispose

Reply:
I agree it's odd, I was following the existing pattern in XmlDependencyRecorder.cs to also end the document when Dispose is called. I think it was implemented in the first place because the end elements need to be added when nothing else is going to be contributed to the document (ie on dispose), but because dgml documents are formatted differently, the entire file needs to be written only after nothing else is being added. I can find a different place in the linker to call `WriteDgmlGraphToFile` and also change it in the XmlDependencyRecorder.
https://github.com/dotnet/linker/blob/5128bae8dfe81adca9ad212ed73e6cc97b33f280/src/linker/Linker/XmlDependencyRecorder.cs#L80-L92

Reply:
Implemented an additional method closes out the document separately from the dispose method, and updated it for both the dgml and xml recorders